### DISCUSSION TITLE
Hardware Mixer/Mute/Solo Controls

### DISCUSSION THREAD

**Discussion by jazzjohn27** [jamulussoftware/jamulus#2106] [Ideas]
A problem with playing on Jamulus is that changing levels, working buttons using a mouse or touchscreen is awkward while playing an instrument. It would be faster and easier to reach over and spin a knob or push a real button.

I have developed a hardware mini-panel with one knob and two buttons for each channel to control the on-screen slider along with the mute and solo buttons.  It will control 4 channels, and it can be cascaded for up to 16 channels. The device puts out serial data containing levels and button states whenever the user makes an adjustment.

I am looking for advice or help tying in to the Jamulus code through a serial port. Basically, I need to control the Jamulus interface via a serial connection.

**Comment by ghost** [comment]
The serial port will be USB. Use a micro-controller with USB on your mini-panel. If your mini-panel and a mouse are used together, then the mouse pointer position can be saved for the GUI items (mixer controls etc) and your mini-panel can move the selected controls with the mouse pointer... for example.

**Comment by pljones** [comment]
I thought we had MIDI control over these now?  `--ctrlmidich` takes a string consisting of a midi channel number and then fader and pan controllers numbers, each with a number of mixing desk channels.  Support for solo and mute are waiting for someone to add them.  The syntax is a bit mind-watering -- there's probably a discussion or issue or pull request at least with more details.  `[MIDI channel];f0*8;p16*8`

**Reply by jazzjohn27** [reply]
A midi interface is good news! I will look into this. I should be able to make code and hardware changes on my circuit to accommodate midi communication.Thank you.

**Reply by pljones** [reply]
Looks like it was https://github.com/jamulussoftware/jamulus/issues/95 and https://github.com/jamulussoftware/jamulus/pull/945 with the details.  You may need to run _without_ the UI to get the MIDI control - that's `--nogui` (or `-n`).

**Reply by jazzjohn27** [reply]
Please clarify... No GUI means I won't see the sliders onscreen?

**Reply by pljones** [reply]
Yes.

I don't know why that decision was taken.

I don't even know if I'm right -- the application doesn't complain on start up.  (Although I might not have dug into it when I was reviewing the start up code, either.)

Hm.  Also, I don't know if it works on Windows...  I'm trying to see if it'll work for me but I've set up my nanoKontrol with a few different banks and working out how to map them to the parameter...  Maybe it does work and I've got the numbers wrong...

**Reply by jazzjohn27** [reply]
this would be for a Pi client (linux)

**Reply by ignotus666** [reply]
It does work with a GUI - I use a midi controller with Jamulus all the time. Details on usage here: 
 https://jamulus.io/wiki/Tips-Tricks-More

**Reply by jazzjohn27** [reply]
Will it work with something like this?
https://www.sweetwater.com/store/detail/XTouchMini--behringer-by-touch-mini

**Reply by ignotus666** [reply]
I should think so - just check if it can send CC data on the buttons too so it can control solo and mute.

**Reply by pljones** [reply]
> note: only available for use with macOS and Linux

Ah, right.  I'll stop trying :).

**Reply by ignotus666** [reply]
> Ah, right. I'll stop trying :).

Oops, yes, it's been quite a lot of time since I wrote that section on midi and I entirely forgot about that!

**Reply by jazzjohn27** [reply]
OK, just ordered it. I'll report back. Thanks to all.

**Reply by WolfganP** [reply]
I remember I tested Jamulus' midi control (and Reaper as well) with this app https://xmmc.de/touchdaw/ , it worked well indeed.

**Reply by ignotus666** [reply]
I just tried out QmidiCtl (Linux) and QmidiNet (Android phone) and was pretty amazed at how well it works. Would be better on a tablet though, my 5" phone is too small to make it useful.

**Comment by jazzjohn27** [comment]
OK, I received the Behringer X Touch Mini but I need a little help. Which driver needs to be installed on the Pi? Thanks.

**Reply by ignotus666** [reply]
I don't have experience with the Pi, but you need to install JACK and QjackCtl to make the connections (does the Pi have a monitor attached?). I assume that's what you're referring to by driver?

**Reply by ignotus666** [reply]
Oh, and a2jmidid.

**Reply by jazzjohn27** [reply]
No monitor. I am using VNC from a Win notebook to see the Jamulus console. By "driver" I guess I mean:  what do I have to do at the command line to get the device to work?

**Reply by ignotus666** [reply]
Just to understand where we are: are you using an audio interface? Do you have JACK, QjackCtl and a2jmidid installed? Do you have any experience using JACK and/or QjackCtl? Doing it all via the terminal is possible but since I always do it via the GUI it's not something I can tell you off the top of my head. If you're using VNC you could equally open and configure QjackCtl, couldn't you?

**Reply by jazzjohn27** [reply]
I installed qjackctl and can run it in the GUI. I see the Jack Audio Connection Kit window in the GUI.

**Reply by jazzjohn27** [reply]
I'll look for a tutorial. I do see the X touch Mini listed under  connections. Thanks.

**Reply by ignotus666** [reply]
Well, now it's just a matter of connecting the X touch to Jamulus in the connections window then.

**Reply by jazzjohn27** [reply]
In Qjackctl, under Connect, in the MIDI tab there is no X Touch Mini. Under the ALSA tab, there is "28:X-TOUCH MINI MINI 1"

There is no Jamulus listed under those two tabs, but it is in  the Audio tab.

**Comment by jazzjohn27** [comment]
![screen3](https://user-images.githubusercontent.com/82913615/140440139-920a061c-2450-430f-81f2-dee047c886f5.jpg)
![screen2](https://user-images.githubusercontent.com/82913615/140440148-8e061ece-3fda-4f48-9751-d0b9a71077b6.jpg)
![Screen1](https://user-images.githubusercontent.com/82913615/140440156-2b4f54c1-3c98-4e38-9e8e-c24e48e5bf70.jpg)

**Comment by ignotus666** [comment]
That's what a2jmidid is for. If you've already installed it, in QjackCtl, click on Setup > Options tab. Check "Execute script after Startup:" and type in the field: `a2jmidid -e &`. Then check "Execute script on shutdown:" and type: `killall a2jmidid` (not sure if the latter is strictly necessary but it won't cause any harm).

This way, you won't have to remember to launch a2jmidid every time. Under the MIDI tab there'll be an "a2j" entry with a dropdown arrow and your X touch should be in there, available to connect to Jamulus.

**Reply by ignotus666** [reply]
As for the Jamulus midi in, see the link I posted above where using midi with Jamulus is explained.

**Reply by jazzjohn27** [reply]
Thank you.  a2jmidid is now installed and the script commands placed in qjackctl.  Restarted qjackctl but do not see the a2j entry in the midi tab. 

 Please repost the link where using midi with Jamulus is explained. I don't see it in your previous posts.

**Reply by ignotus666** [reply]
That's weird - what happens when you type `a2jmidid` in the terminal (with QjackCtl running)?
[Here's](https://jamulus.io/wiki/Tips-Tricks-More) the link (scroll down to the bottom).

**Reply by jazzjohn27** [reply]
Below is what happens when I run a2jmidid.
According to the link, it says "Jamulus must be launched with --ctrlmidich" 
Here's the script that launches the Jambox version of Jamulus:
[Desktop Entry]
Name=Jamulus Start
Type=Application
Comment=Online Jamming
Categories=Application
Path=/home/pi
Exec=/usr/local/bin/jamulus_start.sh
Terminal=true
Icon=jamulus
GenericName=Jamulus

How do I put in     --ctrlmidich '1;f0*8;p16*8;s32*8;m48*8' in that script... right on  the jamulus_start.sh line like this?
Exec=/usr/local/bin/jamulus_start.sh --ctrlmidich '1;f0*8;p16*8;s32*8;m48*8'



pi@jambox:~ $ a2jmidid
Directory "/home/pi/.log" does not exist. Creating...
Directory "/home/pi/.log/a2j" does not exist. Creating...
Directory "/home/pi/.config/a2j" does not exist. Creating...
JACK MIDI <-> ALSA sequencer MIDI bridge, version debian/9-2-0-gc600c10 built on Wed Dec 31 16:00:00 1969
Copyright 2006,2007 Dmitry S. Baikov
Copyright 2007,2008,2009,2011,2012 Nedko Arnaudov

Bridge starting...
Using JACK server 'default'
Hardware ports will not be exported.
Bridge started
Press ctrl-c to stop the bridge
port created: Midi Through [14] (capture): Midi Through Port-0
port created: Midi Through [14] (playback): Midi Through Port-0

**Reply by pljones** [reply]
> /usr/local/bin/jamulus_start.sh

That's a script file, not Jamulus.  You'll need to take a look at what it does.

**Reply by jazzjohn27** [reply]
right... Didn't notice that, Peter. Thanks.

**Comment by jazzjohn27** [comment]
Here's the script jamulus_start.sh


[Script.txt](https://github.com/jamulussoftware/jamulus/files/7491220/Script.txt)

**Reply by ignotus666** [reply]
Edit line 133 in that script to `nice -n -18 jamulus --ctrlmidich '1;f08;p168;s328;m488' $JACKARG` and see what happens.

You didn't say whether launching a2jmidid via the terminal worked to make it appear in the MIDI tab. It's possible it had to be launched the first time like that to create its config directories, and after that it may just work with the command in QjackCtl. I've never used a Pi so I'm in unfamiliar territory...

**Reply by jazzjohn27** [reply]
I do see "a2j" in the midi tab after running a2jmidid from the terminal. Under it is "Midi Through [14] (capture): Midi Through port-0"

**Reply by jazzjohn27** [reply]
Jamulus still only appears in the Audio tab.

**Reply by jazzjohn27** [reply]
I'll try the script edit you suggested and report back. 
Is the goal to see both Jamulus and X Touch Mini in the Midi tab?

**Reply by jazzjohn27** [reply]
OK, now 
a2j/Midi Through [14]capture: Midi Through Port-0  
appears in both windows in the Midi Tab, and 
Jamulus/input midi 
appears in the writable window of the midi tab.

**Reply by ghost** [reply]
I noticed that line 84 has an erroneous 'fi'. This ine can be deleted.

**Reply by ignotus666** [reply]
Ok, we're getting somewhere.

When you launch a2jmidid and it appears in the MIDI tab with 'Midi Through Port-0', is the X touch plugged in and displayed in the ALSA tab? a2jmidid should be displaying all the devices from the ALSA tab. If for some reason the X touch isn't showing up under a2j, a workaround could be this: Connect the X Touch to 'Midi Through Port-0' in the ALSA tab, and 'Midi Through Port-0' under a2j to Jamulus in the MIDI tab. Jamulus should now be able to receive MIDI from the X touch.

**Reply by ignotus666** [reply]
And if the above fails there's yet another thing you can try:

In QjackCtl, click on 'Setup' > 'Settings' tab. On the top right, change 'MIDI Driver' to 'raw'. Restart QjackCtl. Now, your X touch should appear in the MIDI tab, only it might be named "midi_capture_1" under "system". The annoying thing about this is that if you turn your MIDI device off and on again, its name will change to "midi_capture_2", and so on. And any other additional MIDI devices all have this generic name, making it hard to identify them, which is why I prefer to use a2j.

**Reply by jazzjohn27** [reply]
Just so I  follow your connection instructions correctly, I labeled the items in the Midi and Alsa windows. Please tell which letters and numbers to connect. Thank you.
![midi](https://user-images.githubusercontent.com/82913615/140659656-3b64d200-464f-432b-b3f8-99067a0d0a14.png)
![ALSO](https://user-images.githubusercontent.com/82913615/140659659-a97983e5-57bc-476b-9a9e-5cfab40cbf00.png)
s.

**Reply by ignotus666** [reply]
Connect C to 1, and F to 3. I'm not sure if you changed 'MIDI driver' to 'raw', those 2 MIDI ports in 'system' throw me off a bit - but if you did, you could try H to 3. If the latter works you can ditch a2jmidid if you want. What's really strange is that a2jmidid picks up Midi Through but not the X-touch from the ALSA tab...

**Reply by jazzjohn27** [reply]
Tried C-1, H-3. Some numbers show up on Jamulus on the profile boxes at the slider bottoms, e.g. "Jazzjohn" is now "0:Jazzjohn" and "jb3Bp2" is now "1:jb3Bp2" (This is on the JazzGenre/JazzjohnPi2 server) so it seems like some progress was made. The X Touch controls still don't move sliders, though.

Tried the raw driver with the additional connection H-3. Still no movement.

**Reply by ignotus666** [reply]
I was just looking at the MIDI command you're giving to Jamulus - `--ctrlmidich '1;f08;p168;s328;m488'` and it's completely wrong; I hadn't realised earlier. You need to set the MIDI notes you want to use in the X touch first, and establish how many channels (fader strips) you want to be able to control in total. In the following example (similar to in Tips & Tricks), you can control 8 channels (fader + solo + mute in each - I'm leaving out pan because there's no knob for it on the X touch):

`--ctrlmidich '1;f0*8;s32*8;m48*8'`. You were missing the asterisks. In the X touch, in channel 1, you need to set the pot to CC0, button 1 to CC32 and button 2 to CC48. In channel 2, pot to CC1, button 1 to CC33, button 2 to CC49, and so on. So first, set these notes in the X touch, then edit the script to include the right command, and lastly, try the connections suggested above again.

**Reply by jazzjohn27** [reply]
Changed the line in the jamulus_start.sh script to:
nice -n -18 jamulus --ctrlmidich '1;f0*8;s32*8;m48*8' $JACKARG

Made three connections:
F - 3
C - 1
and 
H - 1 (after selecting "raw" for the MIDI Driver in qjackctl)
Restarted qjackctl and a2jmidid.

Still no control of sliders from the X Touch, though the musician profiles on the sliders still show a (channel?) number.

ignotus: I very much appreciate all the time you are spending helping me and I do not want to overcome my welcome, so please feel free to leave this issue which seems overly complicated to me.  Maybe I could try the Jambox people who might have set something like this up before. Thanks much.

**Reply by ignotus666** [reply]
If that's a literal copy/paste of the command it's still wrong - there are asterisks missing. In any event, unless you configure the X Touch to send out the MIDI messages that Jamulus expects (or find out what messages it's sending to configure Jamulus to react to those), it's not going to work. I don't know how to assign MIDI messages in that device, but there must be a way.

Don't worry, I don't want to leave you hanging there with a useless device after egging you on ;)

**Reply by rdica** [reply]
He's not using `style` so his asterisks are italicizing. Look at the first number `8`.

**Reply by ignotus666** [reply]
You're right, hadn't noticed that.

**Reply by jazzjohn27** [reply]
I went to enter a jam session last night only to discover that I could not hear audio in my headphones. So after some tense moments,  I was able to revert back to the original script (and update Jamulus to 3.8.1 during my panic.) 

Is the main problem that X Touch does not show up in the Qjackctl MIDI tab?

**Reply by ignotus666** [reply]
You have several issues:

- Yes, you need the X Touch and Jamulus to be in the same tab so you can connect them together. A workaround is to use Midi Through Port as it appears in both tabs using a2j. That would be connecting F-3 and C-1.
- You need to establish what MIDI messages the X Touch is sending and edit them if necessary. They need to be CC messages. Once they have been established/edited, you can configure Jamulus to listen (react) to those messages. As long as there is a mismatch, it won't work.
- The --ctrlmidich command shouldn't have caused your audio to stop working - AFAIK it is unrelated. However:
- I hadn't done so before but now looking into Jambox it appears that JACK configuration is preferably done via separate config files and not through QjackCtl - changes in the latter seems to overwrite those config files which might be the reason why you lost audio. Since I have no idea how this distribution works your best bet is to head to the Jambox discussions page and explain your situation there. Sorry about not being able to help any further :( . I'm pretty sure you'll be able to figure it out in the end.

**Reply by jazzjohn27** [reply]
Thank you very much for your time. You gave it a good shot, and I learned a few things in the process.
I'll look for Jambox folks who might have experience with MIDI control.
